components/script/dom/window.rs

```diff
@@ -1207,6 +1214,8 @@ impl WindowMethods for Window {
     }
 }
 
+pub struct TaskManagement<T: TaskSource>(pub T, pub TaskCanceller);
```

Comment:
> It'd be possible to unify *_task_source in such a way that we would
have only one method.

Have you considered making this a struct that would completely take over the task management from a `Window`?

Perhaps something like:

```rust
struct TaskManager {
    task_cancellers: DomRefCell<HashMap<TaskSourceName, Arc<AtomicBool>>>,
    task_sources: DomRefCell<HashMap<TaskSourceName, Box<TaskSource>>>
}

impl TaskManager {
    fn task_source_with_canceller(&self, name: TaskSourceName) -> (Box<TaskSource>, TaskCanceller);
    fn task_source(&self, name: TaskSourceName)-> Box<TaskSource>;
}
```

Perhaps it could be initialized and passed to `Window::new` when [a new page loads](https://github.com/servo/servo/blob/391d67fa1d3509cc95ae79dd8f5c0dd33624d9a6/components/script/script_thread.rs#L2514), and basically replace all the task source related methods and fields on `Window`?

> one of the
TaskSource implementations is special:
history_traversal_task_source. Not wanting to over complicate things,
I decided to leave the structure this way.

Perhaps if you were to implement `TaskSource` for `HistoryTraversalTaskSource`(or remove it if others confirm this is a good idea, see my other comment), it could work?

Reply:
I thought of that, but not knowing what to do with `HistoryTraversalTaskSource` made me chose something simpler. Truth being, I'd love to move responsibilities outside of `Window` (it's the most overwhelming file in the source code) :). Your proposal of adding a "fake" implementation for `HistoryTraversalTaskSource` sounds good to me and addresses the concern I had. Will address your feedback and push the new version in the next days.

Thanks for the suggestion.

Reply:
Sounds good, you could also try an actually 'real' implementation, since we might actually need it later for https://github.com/servo/servo/issues/21815

I think basically a copy/paste from any of the other task-source, and adding a new `ScriptThreadEventCategory`, should do the trick...

Reply:
Unfortunately, `TaskSource` can't be made a trait object :(.

I do have the `HistoryTraversalTaskSource` implementation. Should I just do that? Or maybe we can somehow modify `TaskSource` to allow it to make a trait object?

Reply:
I guess I could also make a less elegant...

```
struct TaskManager {
    task_cancellers: DomRefCell<HashMap<TaskSourceName, Arc<AtomicBool>>>,
    *_task_sources: DomRefCell<*TaskSource>,
}
```

Reply:
Thanks for the info. 

Could you please copy/paste the error message from the compiler?

Did you use a `Box<TaskSource>`?

Reply:
Of course! Here it's:

```
error[E0038]: the trait `task_source::TaskSource` cannot be made into an object
   --> components/script/task_manager.rs:119:5
    |
119 |     pub fn task_source(&self, name: TaskSourceName) -> Box<TaskSource> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `task_source::TaskSource` cannot be made into an object
    |
    = note: method `queue_with_canceller` has generic type parameters
    = note: method `queue` has generic type parameters
    = note: the trait cannot contain associated consts like `NAME
```

Reply:
I see, so that's not going to work :)

I haven't found another way to avoid the repetition of methods, have you?

In any case, introducing a `TaskManager` that will lighten up `Window` on methods and fields is probably progress. 

Thanks for pushing to additional commits, I will review them as is unless we come up with a better design...

Reply:
I tried to trick the compiler in a couple naive ways (i.e. using `& mut` instead of `Box`, wrapping the `TaskSource` in a structure that is object safe, etc.). As expected, the compiler outsmarted me.

Another thing I considered was just living with the duplication, but create a couple macros so it's less apparent. I didn't want to go through that route, because I didn't see any macro in the rest of the code and I know some projects are (rightfully) cautious about them. If this were a personal project, though, and the choice depended only on me, I'd probably do that.

Reply:
We do use macros when it makes sense in Servo's codebase.

Reply:
Good to know! Will push (yet another) commit with how it would look like with macros. Thanks!